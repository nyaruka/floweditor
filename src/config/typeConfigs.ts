import AddGroupsForm from '../component/actions/ChangeGroups/AddGroupsForm';
import ChangeGroupsComp from '../component/actions/ChangeGroups/ChangeGroups';
import RemoveGroupsForm from '../component/actions/ChangeGroups/RemoveGroupsForm';
import SendMsgComp from '../component/actions/SendMsg/SendMsg';
import SendMsgForm from '../component/actions/SendMsg/SendMsgForm';
import SetRunResultComp from '../component/actions/SetRunResult/SetRunResult';
import SetRunResultForm from '../component/actions/SetRunResult/SetRunResultForm';
import SetContactFieldComp from '../component/actions/SetContactField/SetContactField';
import SetContactFieldForm from '../component/actions/SetContactField/SetContactFieldForm';
import SendEmailComp from '../component/actions/SendEmail/SendEmail';
import SendEmailForm from '../component/actions/SendEmail/SendEmailForm';
import StartFlowComp from '../component/actions/StartFlow/StartFlow';
import CallWebhookComp from '../component/actions/CallWebhook/CallWebhook';
import GroupsRouter from '../component/routers/GroupsRouter';
import SubflowRouter from '../component/routers/SubflowRouter';
import SwitchRouter from '../component/routers/SwitchRouter';
import WebhookRouter from '../component/routers/WebhookRouter';
import { AnyAction } from '../flowTypes';

/*
Old name	                New name	                Event(s) generated

add_urn	                    add_contact_urn	            contact_urn_added
add_to_group	            add_contact_groups	        contact_groups_added
remove_from_group	        remove_contact_groups	    contact_groups_removed
set_preferred_channel	    set_contact_channel	        contact_channel_changed
update_contact	            set_contact_property	    contact_property_changed
save_contact_field      	set_contact_field	        contact_field_changed
save_flow_result	        set_run_result	            run_result_changed
call_webhook	            call_webhook	            webhook_called
add_label	                add_input_labels	        input_labels_added
reply	                    send_msg	                msg_created
send_email	                send_email	                email_created / email_sent
send_msg	                send_broadcast	            broadcast_created
start_flow	                start_flow	                flow_triggered
start_session	            start_session	            session_triggered
*/

export enum Mode {
    EDITING = 0x1,
    TRANSLATING = 0x2,
    ALL = EDITING | TRANSLATING
}

export interface Type {
    type: string;
    name: string;
    description: string;
    allows(mode: Mode): boolean;
    component?: React.SFC<AnyAction>;
    form?: React.ComponentClass<any>;
    advanced?: Mode;
    aliases?: string[];
}

export interface TypeMap {
    [propName: string]: Type;
}

export type GetTypeConfig = (type: string) => Type;

export function allows(mode: Mode): boolean {
    return (this.advanced & mode) === mode;
}

export const typeConfigList: Type[] = [
    /** Actions */
    {
        type: 'send_msg',
        name: 'Send Message',
        description: 'Send them a message',
        form: SendMsgForm,
        component: SendMsgComp,
        advanced: Mode.EDITING,
        allows
    },
    // { type: 'msg', name: 'Send Message', description: 'Send somebody else a message', form: SendMessageForm, component: SendMessage },
    {
        type: 'add_contact_groups',
        name: 'Add to Group',
        description: 'Add them to a group',
        form: AddGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: 'remove_contact_groups',
        name: 'Remove from Group',
        description: 'Remove them from a group',
        form: RemoveGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: 'set_contact_field',
        name: 'Update Contact',
        description: 'Update the contact',
        form: SetContactFieldForm,
        component: SetContactFieldComp,
        aliases: ['set_contact_property'],
        allows
    },
    {
        type: 'send_email',
        name: 'Send Email',
        description: 'Send an email',
        form: SendEmailForm,
        component: SendEmailComp,
        allows
    },
    {
        type: 'set_run_result',
        name: 'Save Flow Result',
        description: 'Save a result for this flow',
        form: SetRunResultForm,
        component: SetRunResultComp,
        allows
    },
    // {type: 'add_label', name: 'Add Label', description: 'Label the message', component: Missing},
    // {type: 'set_preferred_channel', name: 'Set Preferred Channel', description: 'Set their preferred channel', component: Missing},
    /** Hybrids */
    {
        type: 'call_webhook',
        name: 'Call Webhook',
        description: 'Call a webook',
        form: WebhookRouter,
        component: CallWebhookComp,
        advanced: Mode.EDITING,
        aliases: ['webhook'],
        allows
    },
    {
        type: 'start_flow',
        name: 'Run Flow',
        description: 'Run another flow',
        form: SubflowRouter,
        component: StartFlowComp,
        aliases: ['subflow'],
        allows
    },

    /** Routers */
    {
        type: 'split_by_expression',
        name: 'Split by Expression',
        description: 'Split by a custom expression',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        allows
    },
    {
        type: 'split_by_group',
        name: 'Split by Group Membership',
        description: 'Split by group membership',
        form: GroupsRouter,
        allows
    },
    {
        type: 'wait_for_response',
        name: 'Wait for Response',
        description: 'Wait for them to respond',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        aliases: ['switch'],
        allows
    }
    // {type: 'random', name: 'Random Split', description: 'Split them up randomly', form: RandomRouterForm}
];

export const actionConfigList = typeConfigList.filter(
    ({ type }) =>
        type !== 'wait_for_response' &&
        type !== 'split_by_expression' &&
        type !== 'split_by_group' &&
        type !== 'run_flow' &&
        type !== 'call_webhook'
);

export const typeConfigMap: TypeMap = typeConfigList.reduce((map: TypeMap, typeConfig: Type) => {
    map[typeConfig.type] = typeConfig;
    if (typeConfig.aliases) {
        typeConfig.aliases.forEach((alias: string) => (map[alias] = typeConfig));
    }
    return map;
}, {});

/**
 * Shortcut for constant lookup of type config in type configs map
 * @param {string} type - The type of the type config to return, e.g. 'send_msg'
 * @returns {Object} - The type config found at typeConfigs[type] or -1
 */
export const getTypeConfig = (type: string): Type => typeConfigMap[type];
