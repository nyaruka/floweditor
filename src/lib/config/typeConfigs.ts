import { AnyAction } from '../flowTypes';
import { Language } from '../component/LanguageSelector';
import ChangeGroupComp from '../component/actions/ChangeGroup/ChangeGroup';
import AddGroupForm from '../component/actions/ChangeGroup/AddGroupForm';
import RemoveGroupForm from '../component/actions/ChangeGroup/RemoveGroupForm';
import SaveFlowResultComp from '../component/actions/SaveFlowResult/SaveFlowResult';
import SaveFlowResultForm from '../component/actions/SaveFlowResult/SaveFlowResultForm';
import SaveToContactComp from '../component/actions/SaveToContact/SaveToContact';
import SaveToContactForm from '../component/actions/SaveToContact/SaveToContactForm';
import ReplyComp from '../component/actions/Reply/Reply';
import ReplyForm from '../component/actions/Reply/ReplyForm';
import WebhookComp from '../component/actions/Webhook/Webhook';
import StartFlowComp from '../component/actions/StartFlow/StartFlow';
import SendEmailComp from '../component/actions/SendEmail/SendEmail';
import SendEmailForm from '../component/actions/SendEmail/SendEmailForm';
import SwitchRouterForm from '../component/routers/SwitchRouter';
import SubflowRouterForm from '../component/routers/SubflowRouter';
import WebhookRouterForm from '../component/routers/WebhookRouter';

export enum Mode {
    EDITING = 0x1,
    TRANSLATING = 0x2,
    ALL = EDITING | TRANSLATING
}

export interface Type {
    type: string;
    name: string;
    description: string;
    allows(mode: Mode): boolean;
    component?: React.SFC<AnyAction>;
    form?: React.ComponentClass<any>;
    advanced?: Mode;
    aliases?: string[];
}

export interface TypeMap {
    [propName: string]: Type;
}

export type GetTypeConfig = (type: string) => Type;

export const REPLY_ACTION_TYPE = 'reply';
export const REPLY_ACTION_NAME = 'Send Message';
export const REPLY_ACTION_DESC = 'Send them a message';

export const ADD_GROUP_ACTION_TYPE = 'add_to_group';
export const ADD_GROUP_ACTION_NAME = 'Add to Group';
export const ADD_GROUP_ACTION_DESC = 'Add them to a group';

export const REMOVE_GROUP_ACTION_TYPE = 'remove_from_group';
export const REMOVE_GROUP_ACTION_NAME = 'Remove from Group';
export const REMOVE_GROUP_ACTION_DESC = 'Remove them from a group';

export const typeConfigList: Type[] = [
    /** Actions */
    {
        type: REPLY_ACTION_TYPE,
        name: REPLY_ACTION_NAME,
        description: REPLY_ACTION_DESC,
        form: ReplyForm,
        component: ReplyComp,
        advanced: Mode.EDITING,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    // { type: 'msg', name: 'Send Message', description: 'Send somebody else a message', form: SendMessageForm, component: SendMessage },
    {
        type: ADD_GROUP_ACTION_TYPE,
        name: ADD_GROUP_ACTION_NAME,
        description: ADD_GROUP_ACTION_DESC,
        form: AddGroupForm,
        component: ChangeGroupComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: REMOVE_GROUP_ACTION_TYPE,
        name: REMOVE_GROUP_ACTION_NAME,
        description: REMOVE_GROUP_ACTION_DESC,
        form: RemoveGroupForm,
        component: ChangeGroupComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'save_contact_field',
        name: 'Update Contact',
        description: 'Update the contact',
        form: SaveToContactForm,
        component: SaveToContactComp,
        aliases: ['update_contact'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'send_email',
        name: 'Send Email',
        description: 'Send an email',
        form: SendEmailForm,
        component: SendEmailComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'save_flow_result',
        name: 'Save Flow Result',
        description: 'Save a result for this flow',
        form: SaveFlowResultForm,
        component: SaveFlowResultComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    // {type: 'add_label', name: 'Add Label', description: 'Label the message', component: Missing},
    // {type: 'set_preferred_channel', name: 'Set Preferred Channel', description: 'Set their preferred channel', component: Missing},
    /** Hybrids */
    {
        type: 'call_webhook',
        name: 'Call Webhook',
        description: 'Call a webook',
        form: WebhookRouterForm,
        component: WebhookComp,
        advanced: Mode.EDITING,
        aliases: ['webhook'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'start_flow',
        name: 'Run Flow',
        description: 'Run another flow',
        form: SubflowRouterForm,
        component: StartFlowComp,
        aliases: ['subflow'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },

    /** Routers */
    {
        type: 'expression',
        name: 'Split by Expression',
        description: 'Split by a custom expression',
        form: SwitchRouterForm,
        advanced: Mode.TRANSLATING,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'group',
        name: 'Split by Group Membership',
        description: 'Split by group membership',
        form: SwitchRouterForm,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'wait_for_response',
        name: 'Wait for Response',
        description: 'Wait for them to respond',
        form: SwitchRouterForm,
        advanced: Mode.TRANSLATING,
        aliases: ['switch'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    }
    // {type: 'random', name: 'Random Split', description: 'Split them up randomly', form: RandomRouterForm}
];

export const actionConfigList: Type[] = typeConfigList.filter(
    ({ name }) =>
        name !== 'Wait for Response' &&
        name !== 'Split by Expression' &&
        name !== 'Run Flow' &&
        name !== 'Call Webhook'
);

export const typeConfigMap: TypeMap = typeConfigList.reduce(
    (map: TypeMap, typeConfig: Type) => {
        map[typeConfig.type] = typeConfig;
        if (typeConfig.aliases) {
            typeConfig.aliases.forEach(
                (alias: string) => (map[alias] = typeConfig)
            );
        }
        return map;
    },
    {}
);

/**
 * Shortcut for constant lookup of type config in type configs map
 * @param {string} type - The type of the type config to return, e.g. 'reply'
 * @returns {Object} - The type config found at typeConfigs[type] or -1
 */
export const getTypeConfig = (type: string): Type => typeConfigMap[type];
