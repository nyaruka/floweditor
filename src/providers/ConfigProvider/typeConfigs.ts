import { AnyAction } from '../../flowTypes';
import { Language } from '../../components/LanguageSelector';
import ChangeGroupComp from '../../components/actions/ChangeGroup/ChangeGroup';
import AddGroupForm from '../../components/actions/ChangeGroup/AddGroupForm';
import RemoveGroupForm from '../../components/actions/ChangeGroup/RemoveGroupForm';
import SaveFlowResultComp from '../../components/actions/SaveFlowResult/SaveFlowResult';
import SaveFlowResultForm from '../../components/actions/SaveFlowResult/SaveFlowResultForm';
import SaveToContactComp from '../../components/actions/SaveToContact/SaveToContact';
import SaveToContactForm from '../../components/actions/SaveToContact/SaveToContactForm';
import ReplyComp from '../../components/actions/Reply/Reply';
import ReplyForm from '../../components/actions/Reply/ReplyForm';
import WebhookComp from '../../components/actions/Webhook/Webhook';
import StartFlowComp from '../../components/actions/StartFlow/StartFlow';
import SendEmailComp from '../../components/actions/SendEmail/SendEmail';
import SendEmailForm from '../../components/actions/SendEmail/SendEmailForm';
import SwitchRouterForm from '../../components/routers/SwitchRouter';
import SubflowRouterForm from '../../components/routers/SubflowRouter';
import WebhookRouterForm from '../../components/routers/WebhookRouter';

export enum Mode {
    EDITING = 0x1,
    TRANSLATING = 0x2,
    ALL = EDITING | TRANSLATING
}

export interface Type {
    type: string;
    name: string;
    description: string;
    allows(mode: Mode): boolean;
    component?: React.SFC<AnyAction>;
    form?: React.ComponentClass<any>;
    advanced?: Mode;
    aliases?: string[];
}

export interface TypeMap {
    [propName: string]: Type;
}

export type GetTypeConfig = (type: string) => Type;

export const typeConfigList: Type[] = [
    /** Actions */
    {
        type: 'reply',
        name: 'Send Message',
        description: 'Send them a message',
        form: ReplyForm,
        component: ReplyComp,
        advanced: Mode.EDITING,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    // { type: 'msg', name: 'Send Message', description: 'Send somebody else a message', form: SendMessageForm, component: SendMessage },
    {
        type: 'add_to_group',
        name: 'Add to Group',
        description: 'Add them to a group',
        form: AddGroupForm,
        component: ChangeGroupComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'remove_from_group',
        name: 'Remove from Group',
        description: 'Remove them from a group',
        form: RemoveGroupForm,
        component: ChangeGroupComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'save_contact_field',
        name: 'Update Contact',
        description: 'Update the contact',
        form: SaveToContactForm,
        component: SaveToContactComp,
        aliases: ['update_contact'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'send_email',
        name: 'Send Email',
        description: 'Send an email',
        form: SendEmailForm,
        component: SendEmailComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'save_flow_result',
        name: 'Save Flow Result',
        description: 'Save a result for this flow',
        form: SaveFlowResultForm,
        component: SaveFlowResultComp,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    // {type: 'add_label', name: 'Add Label', description: 'Label the message', component: Missing},
    // {type: 'set_preferred_channel', name: 'Set Preferred Channel', description: 'Set their preferred channel', component: Missing},
    /** Hybrids */
    {
        type: 'call_webhook',
        name: 'Call Webhook',
        description: 'Call a webook',
        form: WebhookRouterForm,
        component: WebhookComp,
        advanced: Mode.EDITING,
        aliases: ['webhook'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'start_flow',
        name: 'Run Flow',
        description: 'Run another flow',
        form: SubflowRouterForm,
        component: StartFlowComp,
        aliases: ['subflow'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },

    /** Routers */
    {
        type: 'expression',
        name: 'Split by Expression',
        description: 'Split by a custom expression',
        form: SwitchRouterForm,
        advanced: Mode.TRANSLATING,
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    },
    {
        type: 'wait_for_response',
        name: 'Wait for Response',
        description: 'Wait for them to respond',
        form: SwitchRouterForm,
        advanced: Mode.TRANSLATING,
        aliases: ['switch'],
        allows(mode: Mode): boolean {
            return (this.advanced & mode) === mode;
        }
    }
    // {type: 'random', name: 'Random Split', description: 'Split them up randomly', form: RandomRouterForm}
];

export const actionConfigList: Type[] = typeConfigList.filter(
    ({ name }) =>
        name !== 'Wait for Response' &&
        name !== 'Split by Expression' &&
        name !== 'Run Flow' &&
        name !== 'Call Webhook'
);

export const typeConfigMap: TypeMap = typeConfigList.reduce((map: TypeMap, typeConfig: Type) => {
    map[typeConfig.type] = typeConfig;
    if (typeConfig.aliases) {
        typeConfig.aliases.forEach((alias: string) => (map[alias] = typeConfig));
    }
    return map;
}, {});

/**
 * Shortcut for constant lookup of type config in type configs map
 * @param {string} type - The type of the type config to return, e.g. 'reply'
 * @returns {Object} - The type config found at typeConfigs[type] or -1
 */
export const getTypeConfig = (type: string): Type => typeConfigMap[type];
