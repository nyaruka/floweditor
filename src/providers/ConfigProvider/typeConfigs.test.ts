import { typeConfigList, actionConfigList, typeConfigMap, getTypeConfig } from './typeConfigs';

describe('Providers: ConfigProvider: typeConfigs', () => {
    it('should provide type configs', () => {
        expect(typeConfigList).toMatchSnapshot();
        expect(actionConfigList).toMatchSnapshot();
        expect(typeConfigMap).toMatchSnapshot();

        Object.keys(typeConfigMap).forEach(key =>
            expect(getTypeConfig(key)).toEqual(typeConfigMap[key])
        );
    });
});
