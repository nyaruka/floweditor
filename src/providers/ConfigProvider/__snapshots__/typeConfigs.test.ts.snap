// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Providers: ConfigProvider: typeConfigs should provide type configs 1`] = `
Array [
  Object {
    "advanced": 1,
    "allows": [Function],
    "component": [Function],
    "description": "Send them a message",
    "form": [Function],
    "name": "Send Message",
    "type": "reply",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Add them to a group",
    "form": [Function],
    "name": "Add to Group",
    "type": "add_to_group",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Remove them from a group",
    "form": [Function],
    "name": "Remove from Group",
    "type": "remove_from_group",
  },
  Object {
    "aliases": Array [
      "update_contact",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "name": "Update Contact",
    "type": "save_contact_field",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Send an email",
    "form": [Function],
    "name": "Send Email",
    "type": "send_email",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Save a result for this flow",
    "form": [Function],
    "name": "Save Flow Result",
    "type": "save_flow_result",
  },
  Object {
    "advanced": 1,
    "aliases": Array [
      "webhook",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Call a webook",
    "form": [Function],
    "name": "Call Webhook",
    "type": "call_webhook",
  },
  Object {
    "aliases": Array [
      "subflow",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Run another flow",
    "form": [Function],
    "name": "Run Flow",
    "type": "start_flow",
  },
  Object {
    "advanced": 2,
    "allows": [Function],
    "description": "Split by a custom expression",
    "form": [Function],
    "name": "Split by Expression",
    "type": "expression",
  },
  Object {
    "allows": [Function],
    "description": "Split by group membership",
    "form": [Function],
    "name": "Split by Group Membership",
    "type": "group",
  },
  Object {
    "advanced": 2,
    "aliases": Array [
      "switch",
    ],
    "allows": [Function],
    "description": "Wait for them to respond",
    "form": [Function],
    "name": "Wait for Response",
    "type": "wait_for_response",
  },
]
`;

exports[`Providers: ConfigProvider: typeConfigs should provide type configs 2`] = `
Array [
  Object {
    "advanced": 1,
    "allows": [Function],
    "component": [Function],
    "description": "Send them a message",
    "form": [Function],
    "name": "Send Message",
    "type": "reply",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Add them to a group",
    "form": [Function],
    "name": "Add to Group",
    "type": "add_to_group",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Remove them from a group",
    "form": [Function],
    "name": "Remove from Group",
    "type": "remove_from_group",
  },
  Object {
    "aliases": Array [
      "update_contact",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "name": "Update Contact",
    "type": "save_contact_field",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Send an email",
    "form": [Function],
    "name": "Send Email",
    "type": "send_email",
  },
  Object {
    "allows": [Function],
    "component": [Function],
    "description": "Save a result for this flow",
    "form": [Function],
    "name": "Save Flow Result",
    "type": "save_flow_result",
  },
  Object {
    "allows": [Function],
    "description": "Split by group membership",
    "form": [Function],
    "name": "Split by Group Membership",
    "type": "group",
  },
]
`;

exports[`Providers: ConfigProvider: typeConfigs should provide type configs 3`] = `
Object {
  "add_to_group": Object {
    "allows": [Function],
    "component": [Function],
    "description": "Add them to a group",
    "form": [Function],
    "name": "Add to Group",
    "type": "add_to_group",
  },
  "call_webhook": Object {
    "advanced": 1,
    "aliases": Array [
      "webhook",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Call a webook",
    "form": [Function],
    "name": "Call Webhook",
    "type": "call_webhook",
  },
  "expression": Object {
    "advanced": 2,
    "allows": [Function],
    "description": "Split by a custom expression",
    "form": [Function],
    "name": "Split by Expression",
    "type": "expression",
  },
  "group": Object {
    "allows": [Function],
    "description": "Split by group membership",
    "form": [Function],
    "name": "Split by Group Membership",
    "type": "group",
  },
  "remove_from_group": Object {
    "allows": [Function],
    "component": [Function],
    "description": "Remove them from a group",
    "form": [Function],
    "name": "Remove from Group",
    "type": "remove_from_group",
  },
  "reply": Object {
    "advanced": 1,
    "allows": [Function],
    "component": [Function],
    "description": "Send them a message",
    "form": [Function],
    "name": "Send Message",
    "type": "reply",
  },
  "save_contact_field": Object {
    "aliases": Array [
      "update_contact",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "name": "Update Contact",
    "type": "save_contact_field",
  },
  "save_flow_result": Object {
    "allows": [Function],
    "component": [Function],
    "description": "Save a result for this flow",
    "form": [Function],
    "name": "Save Flow Result",
    "type": "save_flow_result",
  },
  "send_email": Object {
    "allows": [Function],
    "component": [Function],
    "description": "Send an email",
    "form": [Function],
    "name": "Send Email",
    "type": "send_email",
  },
  "start_flow": Object {
    "aliases": Array [
      "subflow",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Run another flow",
    "form": [Function],
    "name": "Run Flow",
    "type": "start_flow",
  },
  "subflow": Object {
    "aliases": Array [
      "subflow",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Run another flow",
    "form": [Function],
    "name": "Run Flow",
    "type": "start_flow",
  },
  "switch": Object {
    "advanced": 2,
    "aliases": Array [
      "switch",
    ],
    "allows": [Function],
    "description": "Wait for them to respond",
    "form": [Function],
    "name": "Wait for Response",
    "type": "wait_for_response",
  },
  "update_contact": Object {
    "aliases": Array [
      "update_contact",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "name": "Update Contact",
    "type": "save_contact_field",
  },
  "wait_for_response": Object {
    "advanced": 2,
    "aliases": Array [
      "switch",
    ],
    "allows": [Function],
    "description": "Wait for them to respond",
    "form": [Function],
    "name": "Wait for Response",
    "type": "wait_for_response",
  },
  "webhook": Object {
    "advanced": 1,
    "aliases": Array [
      "webhook",
    ],
    "allows": [Function],
    "component": [Function],
    "description": "Call a webook",
    "form": [Function],
    "name": "Call Webhook",
    "type": "call_webhook",
  },
}
`;
